File: correct-crypto/python/aead_encrypt.py
#!/usr/bin/env python3
"""Correct authenticated encryption with AES-256-GCM (AEAD).

A fresh random 96-bit nonce is generated per message and never reused; GCM
provides both confidentiality and integrity, and the associated data is
authenticated but not encrypted.
"""
import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def encrypt(key: bytes, plaintext: bytes, aad: bytes = b"") -> bytes:
    nonce = os.urandom(12)                       # unique per message
    ciphertext = AESGCM(key).encrypt(nonce, plaintext, aad)
    return nonce + ciphertext                    # prepend nonce for the receiver


def decrypt(key: bytes, blob: bytes, aad: bytes = b"") -> bytes:
    nonce, ciphertext = blob[:12], blob[12:]
    return AESGCM(key).decrypt(nonce, ciphertext, aad)


if __name__ == "__main__":
    key = AESGCM.generate_key(bit_length=256)
    blob = encrypt(key, b"correct and secure", b"header")
    assert decrypt(key, blob, b"header") == b"correct and secure"
    print("AES-256-GCM round-trip OK")

File: correct-crypto/ruby/aead_encrypt.rb
# frozen_string_literal: true
# Correct AES-256-GCM authenticated encryption with a random IV (OpenSSL).
require 'openssl'

def encrypt(key, plaintext, aad)
  cipher = OpenSSL::Cipher.new('aes-256-gcm')
  cipher.encrypt
  cipher.key = key
  iv = cipher.random_iv                 # fresh random 96-bit nonce
  cipher.auth_data = aad
  ciphertext = cipher.update(plaintext) + cipher.final
  iv + cipher.auth_tag + ciphertext     # 16-byte GCM tag
end

def decrypt(key, blob, aad)
  cipher = OpenSSL::Cipher.new('aes-256-gcm')
  cipher.decrypt
  cipher.key = key
  cipher.iv = blob[0, 12]
  cipher.auth_tag = blob[12, 16]
  cipher.auth_data = aad
  cipher.update(blob[28..]) + cipher.final
end
